core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +357,215 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
+   * 1. Messages for compacted topics must have keys
+   * 2. When magic value = 1, inner messages of a compressed message set must have monotonically increasing offsets
+   *    starting from 0.
+   * 3. When magic value = 1, validate and maybe overwrite timestamps of messages.
+   *
+   * This method will convert the messages based on the following scenarios:
+   * A. Magic value of a message = 0 and messageFormatVersion is above or equals to 0.10.0-DV0
+   * B. Magic value of a message = 1 and messageFormatVersion is lower than 0.10.0-DV0
+   *
+   * If no format conversion or value overwriting is required for messages, this method will perform in-place
+   * operations and avoids re-compression.
    */
   private[kafka] def validateMessagesAndAssignOffsets(offsetCounter: AtomicLong,
+                                                      now: Long = System.currentTimeMillis(),
                                                       sourceCodec: CompressionCodec,
                                                       targetCodec: CompressionCodec,
-                                                      compactedTopic: Boolean = false): ByteBufferMessageSet = {
-    if(sourceCodec == NoCompressionCodec && targetCodec == NoCompressionCodec) {
-      // do in-place validation and offset assignment
-      var messagePosition = 0
-      buffer.mark()
-      while(messagePosition < sizeInBytes - MessageSet.LogOverhead) {
-        buffer.position(messagePosition)
-        buffer.putLong(offsetCounter.getAndIncrement())
-        val messageSize = buffer.getInt()
-        val positionAfterKeySize = buffer.position + Message.KeySizeOffset + Message.KeySizeLength
-        if (compactedTopic && positionAfterKeySize < sizeInBytes) {
-          buffer.position(buffer.position() + Message.KeySizeOffset)
-          val keySize = buffer.getInt()
-          if (keySize <= 0) {
-            buffer.reset()
-            throw new InvalidMessageException("Compacted topic cannot accept message without key.")
-          }
-        }
-        messagePosition += MessageSet.LogOverhead + messageSize
+                                                      compactedTopic: Boolean = false,
+                                                      messageFormatVersion: ApiVersion = ApiVersion.latestVersion,
+                                                      messageTimestampType: TimestampType,
+                                                      messageTimestampDiffMaxMs: Long): ByteBufferMessageSet = {
+    val magicValueToUse = if (messageFormatVersion.onOrAfter(KAFKA_0_10_0_DV0)) Message.MagicValue_V1 else Message.MagicValue_V0
+    if (sourceCodec == NoCompressionCodec && targetCodec == NoCompressionCodec) {
+      // check the magic value
+      if (!hasMagicValue(magicValueToUse)) {
+        // Message format conversion
+        convertNonCompressedMessages(offsetCounter, compactedTopic, now, messageTimestampType, magicValueToUse)
```

Comment:
The changes in this class is a bit complicated, makes the code a bit hard to read. This and the following are some suggestions on simplification.

Do we need convertNonCompressedMessages()? Since we can't do things in place, could we just use the path that deals with recompression to handle it? This will save some duplicated code.